l1-contracts/contracts/chain-registrar/ChainRegistrar.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+import {IBridgehub} from "../bridgehub/IBridgehub.sol";
+import {PubdataPricingMode} from "../state-transition/chain-deps/ZkSyncHyperchainStorage.sol";
+import {IStateTransitionManager} from "../state-transition/IStateTransitionManager.sol";
+import {ETH_TOKEN_ADDRESS} from "../common/Config.sol";
+import {IERC20} from "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/access/Ownable2StepUpgradeable.sol";
+import {IGetters} from "../state-transition/chain-interfaces/IGetters.sol";
+import {ReentrancyGuard} from "../common/ReentrancyGuard.sol";
+import {SafeERC20} from "@openzeppelin/contracts-v4/token/ERC20/utils/SafeERC20.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev ChainRegistrar serves as the main point for chain registration.
+contract ChainRegistrar is Ownable2StepUpgradeable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+    /// @notice Address that will be used for deploying l2 contracts
+    address public l2Deployer;
+    /// ZKsync Bridgehub
+    IBridgehub public bridgehub;
+
+    /// Chains that has been successfully deployed
+    mapping(address => mapping(uint256 => bool)) public deployedChains;
+    /// Proposal for chain registration
+    mapping(address => mapping(uint256 => ChainConfig)) public proposedChains;
+
+    error ProposalNotFound();
+    error BaseTokenTransferFailed();
+    error ChainIsAlreadyDeployed();
+    error ChainIsNotYetDeployed();
+    error BridgeIsNotRegistered();
+
+    /// @notice new chain is deployed
+    event NewChainDeployed(uint256 indexed chainId, address author, address diamondProxy, address chainAdmin);
+
+    /// @notice new chain is proposed to register
+    event NewChainRegistrationProposal(uint256 indexed chainId, address author);
+
+    /// @notice Shared bridge is registered on l2
+    event SharedBridgeRegistered(uint256 indexed chainId, address l2Address);
+
+    /// @notice L2 Deployer has changed
+    event L2DeployerChanged(address newDeployer);
+
+    struct BaseToken {
+        address tokenAddress;
+        address tokenMultiplierSetter;
+        uint128 gasPriceMultiplierNominator;
+        uint128 gasPriceMultiplierDenominator;
+    }
+
+    struct ChainConfig {
```

Comment:
![Uploading image.png…]()

Reply:
sorry, image is broken =(

Reply:
my bad
<img width="1349" alt="image" src="https://github.com/user-attachments/assets/6f49cb50-8035-414a-a0ff-b516b1a254b5">

Reply:
I've tried my best. but it's not perfectly packable